File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        int prefix = 0;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        Map<Integer, ListNode> map = new HashMap<>();
        map.put(0, dummy);
        for(ListNode curr = dummy; curr != null; curr = curr.next) {
            prefix += curr.val;
            map.put(prefix, curr);
        }
        prefix = 0;
        for(ListNode curr = dummy; curr != null; curr = curr.next) {
            prefix += curr.val;
            curr.next = map.get(prefix).next;
        }
        return dummy.next;
    }
}

File: 1171-remove-zero-sum-consecutive-nodes-from-linked-list/1171-remove-zero-sum-consecutive-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeZeroSumSublists(self, head: Optional[ListNode]) -> Optional[ListNode]:
        map = {}
        dummy = ListNode(0)
        dummy.next = head
        curr = dummy
        prefix = 0
        while curr:
            prefix += curr.val
            map[prefix] = curr
            curr = curr.next
        
        prefix = 0
        curr = dummy
        while curr:
            prefix += curr.val
            
            curr.next = map[prefix].next
            
            curr = curr.next
        return dummy.next